datadog/serializer.go

```diff
@@ -0,0 +1,180 @@
+package datadog
+
+import (
+	"bytes"
+	"io"
+	"log"
+	"math"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/segmentio/stats/v4"
+)
+
+// Datagram format: https://docs.datadoghq.com/developers/dogstatsd/datagram_shell
+
+type serializer struct {
+	conn             net.Conn
+	bufferSize       int
+	filters          map[string]struct{}
+	distPrefixes     []string
+	useDistributions bool
+}
+
+func (s *serializer) Write(b []byte) (int, error) {
+	if s.conn == nil {
+		return 0, io.ErrClosedPipe
+	}
+
+	if len(b) <= s.bufferSize {
+		return s.conn.Write(b)
+	}
+
+	// When the serialized metrics are larger than the configured socket buffer
+	// size we split them on '\n' characters.
+	var n int
+
+	for len(b) != 0 {
+		var splitIndex int
+
+		for splitIndex != len(b) {
+			i := bytes.IndexByte(b[splitIndex:], '\n')
+			if i < 0 {
+				panic("stats/datadog: metrics are not formatted for the dogstatsd protocol")
+			}
+			if (i + splitIndex) >= s.bufferSize {
+				if splitIndex == 0 {
+					log.Printf("stats/datadog: metric of length %d B doesn't fit in the socket buffer of size %d B: %s", i+1, s.bufferSize, string(b))
+					b = b[i+1:]
+					continue
+				}
+				break
+			}
+			splitIndex += i + 1
+		}
+
+		c, err := s.conn.Write(b[:splitIndex])
+		if err != nil {
+			return n + c, err
+		}
+
+		n += c
+		b = b[splitIndex:]
+	}
+
+	return n, nil
+}
+
+func (s *serializer) close() {
+	if s.conn != nil {
+		s.conn.Close()
+	}
+}
+
+func (s *serializer) AppendMeasures(b []byte, _ time.Time, measures ...stats.Measure) []byte {
+	for _, m := range measures {
+		b = s.AppendMeasure(b, m)
+	}
+	return b
+}
+
+// AppendMeasure is a formatting routine to append the dogstatsd protocol
+// representation of a measure to a memory buffer.
+// Tags listed in the s.filters are removed. (some tags may not be suitable for submission to DataDog)
+// Histogram metrics will be sent as distribution type if the metric name matches s.distPrefixes
+func (s *serializer) AppendMeasure(b []byte, m stats.Measure) []byte {
```

Comment:
Might be nice to link the doc reference here as well.

Reply:
Will do!